### DISCUSSION TITLE
Shall we prepare a new release soon?

### DISCUSSION THREAD

**Discussion by rubenporras** [eclipse/lsp4e#523] [General]
I think that we have made a lot of improvements since the last release, so we should try to have new release soon, making it before the SimRel 2022-03 would be nice.

Shall we prepare have new release soon, like for example by the end of the week?

**Comment by mickaelistria** [comment]
>  making it before the SimRel 2022-03 would be nice.

That is unfortunately too late. SimRel is supposed to be frozen on the 10th and API changes are usually avoided in RC phase; pushing LSP4E in now will cause some consumers to fail without any opportunity to safely and quietly remediate their failure before release.

But indeed, if more of the pending major tasks are complete, and if we are happy of newer API (aren't we?), then it would be good to release soon. To not introduce confusion, I think it would be better to do it *after* simrel.

**Comment by rubenporras** [comment]
It is fine by me to do it _after_ simrel.

Yes, I am quite happy with the new API. There is a minor issue that bothers me, and that is that `collectAll`, `computeAll`, `computeFirst`, make a call to org.eclipse.lsp4e.LanguageServers.LanguageServerDocumentExecutor.computeVersion()
and save the computed version on the executor itself, makes the executor keep the state of the last call. That means that in principle there is a race condition if the same executor is reused, because consuming the version is usually part of the `CompletableFuture` chain. @mickaelistria , @ahmedneilhussain , do you think this is something we should/could fix?

**Reply by ahmedneilhussain** [reply]
Yeah, I think the statefulness introduced by versioning support is worth it (I am biased, of course!) but it should be made more explicit. In practice from all of our work on porting code to the new API the impetus to reuse an executor rather than creating a new one does not arise all that often but we should forbid it rather than letting users find out the hard way.

In the (now withdrawn) SingleServerDocumentExecutor PR I added a 'hasRun' flag and made the execution methods throw an `IllegalStateException` if reused. I also added `clone()` and `reset()` methods which you thought were unnecessary but maybe just the internal `hasRun` check would be enough (plus some documentation)?

**Comment by rubenporras** [comment]
the internal hasRun check plus some documentation check would be enough, if we want to keep the statefulness. I would be fine with that. Could you resubmit that part?

**Comment by rubenporras** [comment]
I am happy now with the API.

Besides that, even though there is a new eclipse release today, please, do not introduce dependencies to in before releasing.

**Comment by mickaelistria** [comment]
Let's release now!
